module/data/shared/damage-field.mjs

```diff
@@ -0,0 +1,126 @@
+  /*  Model Configuration                         */
+  /* -------------------------------------------- */
+
+  /** @override */
+  static LOCALIZATION_PREFIXES = ["DND5E.DAMAGE"];
+
+  /* -------------------------------------------- */
+
+  /** @override */
+  static defineSchema() {
+    return {
+      number: new NumberField({ min: 0, integer: true }),
+      denomination: new NumberField({ min: 0, integer: true }),
+      bonus: new FormulaField(),
+      types: new SetField(new StringField()),
+      custom: new SchemaField({
+        enabled: new BooleanField(),
+        formula: new FormulaField()
+      }),
+      scaling: new SchemaField({
+        mode: new StringField(),
+        number: new NumberField({ initial: 1, min: 0, integer: true }),
+        formula: new FormulaField()
+      })
+    };
+  }
+
+  /* -------------------------------------------- */
+  /*  Properties                                  */
+  /* -------------------------------------------- */
+
+  /**
+   * The default damage formula.
+   * @type {string}
+   */
+  get formula() {
+    if ( this.custom.enabled ) return this.custom.formula ?? "";
+    return this._automaticFormula();
+  }
+
+  /* -------------------------------------------- */
+  /*  Helpers                                     */
+  /* -------------------------------------------- */
+
+  /**
+   * Produce the auto-generated formula from the `number`, `denomination`, and `bonus`.
+   * @param {number} [increase=0]  Amount to increase the die count.
+   * @returns {string}
+   * @protected
+   */
+  _automaticFormula(increase=0) {
+    let formula;
+    const number = (this.number ?? 0) + increase;
+    if ( number && this.denomination ) formula = `${number}d${this.denomination}`;
+    if ( this.bonus ) formula = formula ? `${formula} + ${bonus}` : bonus;
+    return formula ?? "";
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Scale the damage by a number of steps using its configured scaling configuration.
+   * @param {number} increase  Number of steps above base damage to scaling.
+   * @returns {string}
+   */
+  scaledFormula(increase) {
+    switch ( this.scaling.mode ) {
+      case "whole": break;
+      case "half": increase = Math.floor(increase * .5); break;
+      default: increase = 0; break;
+    }
+    if ( !increase ) return this.formula;
+
+    // If dice count scaling, increase the count on the first die rolled
+    const dieIncrease = (this.scaling.number ?? 0) * increase;
+    if ( this.custom.enabled ) {
+      formula = this.custom.formula;
+      formula = formula.replace(/^(\d)+d/, (match, number) => `${Number(number) + dieIncrease}d`);
+    } else {
+      formula = this._automaticFormula(dieIncrease);
+    }
+
+    // If custom scaling included, modify to match increase and append for formula
+    if ( this.scaling.formula ) {
+      let roll = new Roll(this.scaling.formula);
+      roll = roll.alter(increase, 0, { multiplyNumeric: true });
```

Comment:
Should we be double-dipping here? Normally when you scale a formula, you either scale the base formula, *or* you apply the scaling formula to it. Here we are doing both.

Reply:
I wanted to include the dice scaling because it is what most damage uses so it easier, but still offer the flexibility of custom scaling formulas. I generally you'd use one or the other, but it didn't seem like there was any reason to prevent both if someone had a damage that added dice and numbers (like if someone wanted to scale Magic Missile, they could put `1` in the dice box and `1` into the custom formula box so it increases by `1d4 + 1` for each scaling step).

Reply:
I think I actually misunderstood what this code was doing, but your answer clarifies it for me, thank you. Looks like we are missing a `let formula;` somewhere though.